### ISSUE TITLE
File attachment upload seems to hang

### ISSUE THREAD

**Issue by kec056** [n3xah/issues#3]
![File upload seems to hang](https://github.com/user-attachments/assets/181dd3be-875e-4568-84db-300167f4d906)

This is the file I used. It's about 40 MB, so maybe the larger size was what did it?
https://www.dropbox.com/scl/fi/1tgg9rqsl2ol19lf1fmco/Rick-Steves-Great-Britain-compressed.pdf?rlkey=fjetack7z0led4ud91n9580mz&st=o8wi2dnk&dl=0

**Comment by brentbahry** [comment]
You should get an error at the top of the input field about the size if it's too big. I also believe the max size is 50mb. If you happen to have any console logs that might be helpful.

**Comment by kec056** [comment]
[n3x file upload console logs.txt](https://github.com/user-attachments/files/16322555/n3x.file.upload.console.logs.txt)

[n3xa file upload network logs.txt](https://github.com/user-attachments/files/16322558/n3xa.file.upload.network.logs.txt)

**Comment by kec056** [comment]
You might be able to change the network logs file type to ".har" if that helps

**Comment by brentbahry** [comment]
thanks for the logs, i'll check it out.

separate topic, would you mind reloading the page and giving me the console logs again? i noticed the websocket stuff we just introduced seems to be failing to connect.

**Comment by kec056** [comment]
Logs on page reload, right up to when I try to upload the 40 MB file
[reconnect console logs.txt](https://github.com/user-attachments/files/16323182/reconnect.console.logs.txt)

**Comment by kec056** [comment]
This is the first error after the upload hangs

```
Uncaught (in promise) Error: Failed to process service request: /service/@n3xa/app-common/ThoughtConversationService/saveFiles, error: 
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12633
    step ServiceClient.js:33
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12537
    fulfilled ServiceClient.js:5
    promise callback*step ServiceClient.js:7
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12531
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12527
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12615
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12593
    step ServiceClient.js:33
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12537
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12531
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12527
    sendRequest ServiceClient.js:59
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12608
    step ServiceClient.js:33
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12537
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12531
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12527
    node_modules vendor.cfc0e85c9f93e8a66e52.js:12578
    getSaveFiles Thought.tsx:251
    __awaiter app.e86c3aa625437e0af264.js:5092
    __awaiter app.e86c3aa625437e0af264.js:5088
    getSaveFiles Thought.tsx:250
    uploadFiles ChatInput.tsx:160
    fulfilled app.e86c3aa625437e0af264.js:2803
    promise callback*step app.e86c3aa625437e0af264.js:2805
    __awaiter app.e86c3aa625437e0af264.js:2806
    __awaiter app.e86c3aa625437e0af264.js:2802
    uploadFiles ChatInput.tsx:113
    handleFileChange ChatInput.tsx:102
    __awaiter app.e86c3aa625437e0af264.js:2806
    __awaiter app.e86c3aa625437e0af264.js:2802
    handleFileChange ChatInput.tsx:100
    React 23
    loadApp Router.js:17
    ts index.ts:3
    Webpack 7
[ServiceClient.js:110](webpack://n3xa/common/node_modules/@proteinjs/service/dist/src/ServiceClient.js)
[socket.io] Connected [useSocket.tsx:28:16](webpack://n3xa/app-ui/src/socket/useSocket.tsx)

​
```

**Comment by brentbahry** [comment]
Cool thanks. Looks like the websocket is connecting fine after reload so that's good for that separate topic.

**Comment by brentbahry** [comment]
Just an update @kec056, unfortunately it looks like the PDF you're trying to chat with the assistant about is too large.

tldr: We should fail more gracefully with a clear error.

More details below.

We have a couple size checks, the 50mb max I mentioned before is a general file upload limit. However, we also estimate the `token` (~word) count of all things we send to the assistant. The reason for that is because of the maximum amount of text that models can process at once (which is measured in tokens). For gpt-4o, which we currently use, that limit is 128K tokens. That includes chat history, files, thoughts, and anything else the assistant is provided as context to generate its response. When testing a smaller PDF than yours (34mb), it clocked in at 198k tokens just for the file.

The reason we can't just look at the mb size of the file in the browser is because the token calculation actually leverages a server-side library provided by OpenAI because the calculation isn't as simple as 1 byte = x tokens. On top of that, unintuitively, we can't just read any file as a UTF-8 string and send it to the assistant. Many common file types (word docs, spreadsheets, pdfs, etc) consist of a bunch of metadata (similar to markdown) that surrounds the content. Current day assistants aren't able to see past the metadata, so we have to parse it (differently for each file type we want to support). And we don't really know what the size (or token count) of the parsed files will be until we parse them.

The general challenge this presents is how to communicate what is too big to the user. The cleanest answer would be to expose token count, but I'm not sure that's something the user should have to understand. We've fallen back on just saying the file size is too big for now.
